cpp/src/comms/mpi/comms_mpi.hpp

```diff
@@ -0,0 +1,256 @@
+    if (std::is_signed<value_t>::value) {
+      if (sizeof(value_t) == 1) {
+        return ncclInt8;
+      }
+      else if (sizeof(value_t) == 4) {
+        return ncclInt32;
+      }
+      else if (sizeof(value_t) == 8) {
+        return ncclInt64;
+      }
+      else {
+        CUGRAPH_FAIL("unsupported type");
+      }
+    }
+    else {
+      if (sizeof(value_t) == 1) {
+        return ncclUint8;
+      }
+      else if (sizeof(value_t) == 4) {
+        return ncclUint32;
+      }
+      else if (sizeof(value_t) == 8) {
+        return ncclUint64;
+      }
+      else {
+        CUGRAPH_FAIL("unsupported type");
+      }
+    }
+  }
+  else if(std::is_same<value_t, float>::value) {
+    return ncclFloat32;
+  }
+  else if(std::is_same<value_t, double>::value) {
+    return ncclFloat64;
+  }
+  else {
+    CUGRAPH_FAIL("unsupported type");
+  }
+}
+
+enum class ReduceOp { SUM, MAX, MIN };
+
+constexpr MPI_Op get_mpi_reduce_op(ReduceOp reduce_op) {
+  if (reduce_op == ReduceOp::SUM) {
+    return MPI_SUM;
+  }
+  else if (reduce_op == ReduceOp::MAX) {
+    return MPI_MAX;
+  }
+  else if (reduce_op == ReduceOp::MIN) {
+    return MPI_MIN;
+  }
+  else {
+    CUGRAPH_FAIL("unsupported type");
+  }
+}
+
+constexpr ncclRedOp_t get_nccl_reduce_op(ReduceOp reduce_op) {
+  if (reduce_op == ReduceOp::SUM) {
+    return ncclSum;
+  }
+  else if (reduce_op == ReduceOp::MAX) {
+    return ncclMax;
+  }
+  else if (reduce_op == ReduceOp::MIN) {
+    return ncclMin;
+  }
+  else {
+    CUGRAPH_FAIL("unsupported type");
+  }
+}
+#endif
+
+// basic info about the snmg env setup
+class Comm 
+{ 
+  private:
+    int _p{0};
+
+    int _mpi_world_rank{0};
+    int _mpi_world_size{0};
+    bool _finalize_mpi{false};
+    bool _finalize_nccl{false};
+
+
+    int _device_id{0};
+    int _device_count{0};
+
+    std::vector<void*> _p_ipc_mems{};
+    std::vector<size_t> _local_ipc_mem_offsets{};
+
+    int _sm_count_per_device{0};
+    int _max_grid_dim_1D{0};
+    int _max_block_dim_1D{0};
+    int _l2_cache_size{0};
+    int _shared_memory_size_per_sm{0};
+
+#if USE_NCCL
+    MPI_Comm _mpi_comm{};
+    ncclComm_t _nccl_comm{};
+ #endif
+   
+  public: 
+    Comm(){};
+    Comm(int p);
+    ~Comm();
+    int get_rank() const { return _mpi_world_rank; }
+    int get_p() const { return _mpi_world_size; }
+    int get_dev() const { return _device_id; }
+    int get_dev_count() const { return _device_count; }
+    int get_sm_count() const { return _sm_count_per_device; }
+    bool is_master() const { return (_mpi_world_rank == 0)? true : false; }
+
+    void barrier();
+
+    template <typename value_t>
+    void allgather (size_t size, value_t* sendbuff, value_t* recvbuff) const;
+
+    template <typename value_t>
+    void allreduce (size_t size, value_t* sendbuff, value_t* recvbuff, ReduceOp reduce_op) const;
+
+};
+
+template <typename value_t>
+void Comm::allgather (size_t size, value_t* sendbuff, value_t* recvbuff) const {
+#if USE_NCCL

```

Comment:
If we are not providing both NCCL based and CUDA-aware MPI based implementations, better just delete #if USE_NCCL everywhere.

Reply:
I'm not sure. That helps to activate/deactivate MPI dependent code. If/when we set OPG build ON by default, this will be leveraged to get a successful build for all users that don't have MPI installed.

Reply:
I don't think I am following this comment. Here, if USE_NCCL is set to 0, this function just becomes no-op. If we want to enable users to run this without MPI, removing #if USE_NCCL and #endif does no harm (I didn't mean to delete the code in between, if you interpret this as that, then, my bad, I just meant to delete the #if USE_NCCL lines and matching #endif lines, not the NCCL based implementations between those lines).

Reply:
> If we want to enable users to run this without MPI, removing #if USE_NCCL and #endif does no harm

`#include <mpi.h>` won't build if MPI is not installed

Reply:
I see, in that case USE_NCCL might be a misnomer (better add USE_MPI).

https://github.com/rapidsai/cugraph/pull/840/files#diff-4a9c63a50daac99ed3fa518d6887cfd8R91
```
#if USE_NCCL
  MPI_Barrier(MPI_COMM_WORLD);
#endif 
```

The code above also looks problematic. Better not bog down too deep for a temporary code, but int his code, setting USE_NCCL to 1 will not allow users without MPI to build this code.

Reply:
Yea naming is misleading. Currently, MPI and NCCL are interdependent and the `USE_NCCL` macro is controlling it.

Reply:
Fixed naming 7ac19b1